### PR TITLE
Refactor dns handlers

### PR DESCRIPTION
**Pull request by rade**
This takes the refactoring bit of #1334, evolves it (the handler construction is neater here), adds some extra refactors, and an optimisation.

Best reviewed by looking at the individual commits. And the first of those is best view in a good diff viewer, e.g. kompare in whitespace-ignore mode, where it will appear *much* smaller than here on github.

### CODE REVIEW

File: nameserver/dns_test.go [hunk 1]
```diff
@@ -83,17 +83,17 @@ func TestTruncateResponse(t *testing.T) {
...
+		answers := make([]dns.RR, numAnswers)
 		for j := 0; j < numAnswers; j++ {
-			response.Answer[j] = &dns.A{Hdr: header, A: address.Address(j).IP4()}
+			answers[j] = &dns.A{Hdr: header, A: address.Address(j).IP4()}
 		}
 
 		// pick a random max size, truncate response to that, check it
-		maxSize := 512 + rand.Intn(response.Len()-512)
-		truncateResponse(response, maxSize)
+		maxSize := 512 + rand.Intn(2*512)
```
Comment (rade) [inline nameserver/dns_test.go]: This change isn't ideal, but all the alternatives I can think of are worse.

**Comment by tomwilkie** [comment]
If you stick ?w=1 on the end of the url, github will also ignore whitespace.

**Comment by rade** [comment]
> If you stick ?w=1 on the end of the url, github will also ignore whitespace.

Thanks! Looks much better that way.

File: nameserver/dns.go [hunk 2]
```diff
@@ -112,183 +112,171 @@ func (d *DNSServer) errorResponse(r *dns.Msg, code int, w dns.ResponseWriter) {
...
+	for _, server := range h.upstream.Servers {
+		reqCopy := req.Copy()
+		reqCopy.Id = dns.Id()
+		response, _, err := h.client.Exchange(reqCopy, fmt.Sprintf("%s:%s", server, h.upstream.Port))
+		if err != nil || response == nil {
+			h.ns.debugf("error trying %s: %v", server, err)
+			continue
+		}
+		response.Id = req.Id
+		if response.Len() > h.getMaxResponseSize(req) {
```
Comment (tomwilkie) [inline nameserver/dns.go]: missing the -1 check here.

Comment (rade) [inline nameserver/dns.go]: This is meant to be a refactor, not fix any bugs.

**Comment by tomwilkie** [comment]
Whilst you're refactoring, you could move errorResponse onto handler and move it down the file to be with makeResponse.

**Comment by tomwilkie** [comment]
Also should errorResponse use respond too?

**Comment by tomwilkie** [comment]
Otherwise LGTM; I'll fix #1345 and #1332 on top of this now?

**Comment by rade** [comment]
> I'll fix #1345 and #1332 on top of this now?

wait until it's merged.

**Comment by tomwilkie** [comment]
That last change looks good to; merge away.

**Pull request merged**